nano/node/blocking_observer.cpp

```diff
@@ -0,0 +1,49 @@
+#include <nano/node/blocking_observer.hpp>
+#include <nano/node/blockprocessor.hpp>
+
+void nano::blocking_observer::connect (nano::block_processor & block_processor)
+{
+	block_processor.batch_processed.add ([this] (auto const & items) {
+		for (auto const & [result, block] : items)
+		{
+			observe (block, result);
+		}
+	});
+}
+
+void nano::blocking_observer::stop ()
+{
+	nano::lock_guard<nano::mutex> lock{ mutex };
+	stopped = true;
+	blocking.clear (); // ~promise future_error
+}
+
+void nano::blocking_observer::observe (std::shared_ptr<nano::block> block, nano::process_return const & result)
+{
+	nano::lock_guard<nano::mutex> lock{ mutex };
+	auto existing = blocking.find (block);
+	if (existing != blocking.end ())
+	{
+		existing->second.set_value (result);
+		blocking.erase (existing);
+	}
+}
+
+std::future<nano::process_return> nano::blocking_observer::insert (std::shared_ptr<nano::block> block)
+{
+	nano::lock_guard<nano::mutex> lock{ mutex };
+	if (stopped)
+	{
+		std::promise<nano::process_return> promise;
+		return promise.get_future (); // ~promise future_error
+	}
+	auto iterator = blocking.emplace (block, std::promise<nano::process_return>{});
```

Comment:
progress is the default enum value, but wouldn't be better to have it written in the constructor?

Reply:
This is a promise for an enum value rather than the value itself.